---
layout: post
title: 3. Decide if you want to code.
---
<p align="center">
  <img src="https://media.giphy.com/media/26tPnAAJxXTvpLwJy/giphy.gif">
</p>

It is not always necessary to code to analyze large biological datasets. In many cases, graphical user interfaces (GUIs) will allow you to explore and get an intuition for your large datasets quickly. Bulk RNA-Seq users can use Galaxy. Single cell RNA-Seq experiments performed on a 10X Chromium may be able to use Loupe. ATAC-Seq people can use Guava. ChIP-Seq people can use ChIPseeker.

Learning to code, however, will give you three massive advantages: control, reproducibility, and hireability. The more you can tinker with a piece of software, the more flexibility you have to customize the assumptions in the analysis, the exploration of the data, the resulting visualizations. Learning how to write well-organized code will also make it easier for you to say exactly why and how you got a result instead of trying to remember the specific order in which you clicked buttons. Being able to interface with code is also a sought-after job skill that will make you more appealing as a job candidate in both academia and industry.

What languages should you learn? This will depend on your particular bioinformatic specialty. Pretty much every data scientist that uses a computational cluster or Unix (MacOS or Linux) would benefit from learning some basic bash commands for navigating through folders on your computer, peeking at large files, and running scripts. For transcriptomics, I recommend either using R or Python and sticking with that until you feel very fluent in it.

How do you choose which to use? If you are around other coders, use what they use, because it will help your progress tremendously if there are people you can ask for help (more on that later). If you are your lab’s only coder, and doing transcriptomics or Prism-type statistics, I think it’s a toss-up. Python is more flexible and a highly desired skill in a wide array of positions outside academia (although many positions want R). Python is also the preferred method for extremely large datasets (millions of single cell RNA-Seq datapoints!) or for machine learning. R has more widely-vetted ready-made libraries for transcriptomics analysis (although Python is catching up). Both languages will help you do good work and get a job someday. 

If you choose Python, I suggest that you start writing your pipelines in JuPyTer. If you choose R, start writing your pipelines in RStudio using R Notebooks. Both of these formats will allow you to work with code interactively and create neat, formatted, highly annotated versions of your code ready for export into a lab notebook.

Coding isn’t for everyone, and you may hate it. It can be really difficult to tell the difference between frustration and true coding hatred for quite some time. I recommend trying to get through the visualization-producing step of your pipeline before you decide to give up coding forever. It feels great to transform a matrix of numbers into a nice-looking visual! 

Remember - even if you don’t fall in love with coding - having tried it will make it easier to communicate effectively with the people who did.

I am such a big advocate for learning to code, in fact, that I’ll proceed with this outline with the assumption you want to code. 

Next: [Step Four](https://kmuench.github.io/2020/03/18/step-4/) - or return to the [overview](https://kmuench.github.io/2020/03/18/ten-steps-to-bioinf/).
